### DISCUSSION TITLE
How to make python codes friendly to code indexers?

### DISCUSSION THREAD

**Discussion by ImagineBoom** [gem5/gem5#736] [Q&A] [answered]
**The origin of the problem**
I use **vscode**, and I want to be able to get more information about objects when I'm programming, like PowerState..., which I know is in the src/sim directory, but the editor doesn't know it
![image](https://github.com/gem5/gem5/assets/39993563/ded850cb-47bd-4de1-b3e9-af29d8095061)

if I replace "m5.objects" to "sim", build error
![image](https://github.com/gem5/gem5/assets/39993563/bfa8c978-3396-43ac-a290-a64ff571d3a6)
**err log:**
```
terminate called after throwing an instance of 'pybind11::error_already_set'
  what():  ModuleNotFoundError: No module named 'sim'

At:
  src/sim/ClockedObject.py(36): <module>
  <string>(40): exec_module
  <frozen importlib._bootstrap>(703): _load_unlocked
  <frozen importlib._bootstrap>(1006): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1027): _find_and_load
  src/learning_gem5/part2/SimpleCache.py(30): <module>
  <string>(40): exec_module
  <frozen importlib._bootstrap>(703): _load_unlocked
  <frozen importlib._bootstrap>(1006): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1027): _find_and_load
  <string>(1): <module>
  src/python/m5/objects/__init__.py(29): <module>
  <string>(40): exec_module
  <frozen importlib._bootstrap>(703): _load_unlocked
  <frozen importlib._bootstrap>(1006): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1027): _find_and_load
  <frozen importlib._bootstrap>(1050): _gcd_import
  <frozen importlib._bootstrap>(241): _call_with_frames_removed
  <frozen importlib._bootstrap>(992): _find_and_load_unlocked
  <frozen importlib._bootstrap>(1027): _find_and_load
  <frozen importlib._bootstrap>(1050): _gcd_import
  /usr/lib/python3.10/importlib/__init__.py(126): import_module
  build_tools/sim_object_param_struct_cc.py(71): <module>

Aborted
scons: *** [build/X86/python/_m5/param_VncInput.cc] Error 134
scons: building terminated because of errors.
```

**Comment by MeatBoy106** [comment, marked as answer]
The m5 module is not only defined by the shape of the source tree like a regular Python module but rather elaborated by the build system when compiling gem5. In particular, all SimObject's modules are flattened into the m5.objects module.
It starts with the SimObject declarations you find in SConscript files (implemented in `src/SConscript`). From there, it gets a bit convoluted and definitely out of reach of any standard Python code indexer.

In general, gem5 is tough on code indexers as it relies on C++ code generation (SimObject parameters and Ruby), complex introspection and reflection in Python as well as some build time source configuration (e.g., ISA selection).

**Reply by ImagineBoom** [reply]
Thanks, this is where I am puzzled: How to make python codes friendly to code indexers? I have changed the title of the discussion.

**Reply by MeatBoy106** [reply]
I don't know if you can do much better. Indexers are relatively static in the way they analyze code while the state of gem5 modules is defined at runtime, which an indexer can't really grasp.

Maybe an expert Python developer could provide valuable hints but I would bet there is no good (simple and reliable) option there.

**Reply by powerjg** [reply]
I believe we've gotten some stubs created before. See https://github.com/gem5/gem5/pull/307

If you run that script, you should get some *better* indexing, but it won't be perfect (e.g., I think it's missing the parameters and/or parameter types)
